File: android/core/src/main/kotlin/io/github/kegustafsson/driveremote/core/DriveCommand.kt
package io.github.kegustafsson.driveremote.core

/**
 * Port of `sk-plugin/src/pure/driveCommand.ts`, which is itself a deliberate
 * mirror of the firmware's `lib/control_core/drive/drive_command.h`. Three
 * copies of this truth table now exist -- C++, TypeScript and Kotlin -- so the
 * vectors in `DriveCommandTest` are ported one-for-one from
 * `driveCommand.test.ts` and `test_drive_command.cpp`. Change one, change all
 * three, and the tests are what will tell you that you did.
 *
 * Unlike the TypeScript, Kotlin has no string-union type, so the wire
 * representation is carried explicitly on the enum. That is closer to the
 * firmware's own enum + `ToSkString`/`FromSkString` boundary conversion than
 * the TypeScript is.
 */
enum class DrivePosition(val wire: String) {
  FORWARD("forward"),
  NEUTRAL("neutral"),
  REVERSE("reverse"),
}

/**
 * What a *display* may show for a drive. Deliberately a separate type from
 * [DrivePosition], carrying a fourth [UNKNOWN] case: telemetry that cannot be
 * parsed must be shown as unknown, never silently relabelled as neutral.
 *
 * Commanding has no such case -- there the safe answer is always a real
 * position -- which is exactly why these are two types and not one nullable
 * one. AGENTS.md: never present unconfirmable data as live; degrade the
 * indication instead.
 */
enum class DisplayDrivePosition {
  FORWARD,
  NEUTRAL,
  REVERSE,
  UNKNOWN,
}

/**
 * Maps one side's two momentary button states to a commanded position.
 *
 * Both active at once fails to [DrivePosition.NEUTRAL], never to a direction.
 * A single physical switch cannot produce that input, but two independent
 * touch targets pressed by two different fingers can -- and on a phone that is
 * an ordinary thing to do by accident. SAFETY.md drive invariant 1: never
 * assert FORWARD and REVERSE together.
 */
fun fromSwitch(forwardActive: Boolean, reverseActive: Boolean): DrivePosition =
  when {
    forwardActive && reverseActive -> DrivePosition.NEUTRAL
    forwardActive -> DrivePosition.FORWARD
    reverseActive -> DrivePosition.REVERSE
    else -> DrivePosition.NEUTRAL
  }

/**
 * Defensive parse of a drive position arriving FROM Signal K (RX's telemetry,
 * read only for display). Anything unrecognised -- wrong case, wrong type,
 * absent -- reads as [DisplayDrivePosition.UNKNOWN] rather than a guess.
 */
fun parseDisplayPosition(raw: Any?): DisplayDrivePosition =
  when (raw) {
    "forward" -> DisplayDrivePosition.FORWARD
    "neutral" -> DisplayDrivePosition.NEUTRAL
    "reverse" -> DisplayDrivePosition.REVERSE
    else -> DisplayDrivePosition.UNKNOWN
  }

File: sk-plugin/src/pure/driveCommand.ts
// Pure TypeScript, no DOM/React/WebSocket deps -- host-testable in
// isolation, same discipline as lib/control_core/drive_command.h in the
// firmware (this file is a deliberate line-for-line mirror of it; keep the
// two in sync if either changes).
//
// DrivePosition IS the wire representation here (unlike the firmware,
// which keeps an internal enum and converts at the SK boundary via
// ToSkString/FromSkString) -- in TypeScript the string union type serves
// both roles with no loss of safety, since the type system already
// prevents constructing anything outside these three values from
// literals. fromSwitch() below still can't produce an invalid value even
// though there's no separate "boundary conversion" step to do it in.

export type DrivePosition = 'forward' | 'neutral' | 'reverse';

/**
 * Maps a side's two momentary button states to a position. Structurally
 * mirrors control_core::FromSwitch() in drive_command.h, including its
 * fail-to-neutral behavior when both contacts are active at once -- which
 * a single physical switch can never do, but which IS reachable here: two
 * independent touch buttons (forward, reverse) can both be pressed by two
 * different fingers simultaneously. Neutral, not an arbitrary pick, is the
 * only safe answer (SAFETY.md drive invariant 1: never both directions).
 */
export function fromSwitch(
  forwardActive: boolean,
  reverseActive: boolean,
): DrivePosition {
  if (forwardActive && reverseActive) return 'neutral';
  if (forwardActive) return 'forward';
  if (reverseActive) return 'reverse';
  return 'neutral';
}

/**
 * Defensive parse for values arriving FROM Signal K (RX's telemetry, read
 * for the status display) -- distinct from DrivePosition because a
 * display needs to be able to say "I don't actually know," which
 * commanding never does. Mirrors control_core::FromSkString()'s
 * fail-safe-on-garbage philosophy, but for display purposes silently
 * relabeling unparseable telemetry as "neutral" would misrepresent RX's
 * actual state rather than protect anything, so this returns a distinct
 * 'unknown' instead of guessing.
 */
export type DisplayDrivePosition = DrivePosition | 'unknown';

export function parseDisplayPosition(raw: unknown): DisplayDrivePosition {
  if (raw === 'forward' || raw === 'neutral' || raw === 'reverse') return raw;
  return 'unknown';
}
